server/src/main/java/org/elasticsearch/immutablestate/service/ImmutableClusterStateController.java

```diff
@@ -0,0 +1,302 @@
+     * @param immutableStateFilePackage a {@link Package} composite state object to process
+     * @param errorListener a consumer called with IllegalStateException if the content has errors and the
+     *        cluster state cannot be correctly applied, IncompatibleVersionException if the content is stale or
+     *        incompatible with this node {@link Version}, null if successful.
+     */
+    public void process(String namespace, Package immutableStateFilePackage, Consumer<Exception> errorListener) {
+        Map<String, Object> immutableState = immutableStateFilePackage.state;
+        PackageVersion packageVersion = immutableStateFilePackage.metadata;
+
+        LinkedHashSet<String> orderedHandlers;
+        try {
+            orderedHandlers = orderedStateHandlers(immutableState.keySet());
+        } catch (Exception e) {
+            List<String> errors = List.of(e.getMessage());
+            recordErrorState(
+                new ImmutableUpdateErrorState(namespace, packageVersion.version(), errors, ImmutableStateErrorMetadata.ErrorKind.PARSING)
+            );
+            logger.error("Error processing state change request for [{}] with the following errors [{}]", namespace, errors);
+
+            errorListener.accept(new IllegalStateException("Error processing state change request for " + namespace, e));
+            return;
+        }
+
+        ClusterState state = clusterService.state();
+        ImmutableStateMetadata existingMetadata = state.metadata().immutableStateMetadata().get(namespace);
+        if (checkMetadataVersion(existingMetadata, packageVersion, errorListener) == false) {
+            return;
+        }
+
+        // Do we need to retry this, or it retries automatically?
+        clusterService.submitStateUpdateTask(
+            "immutable cluster state [" + namespace + "]",
+            new ImmutableStateUpdateStateTask(
+                namespace,
+                immutableStateFilePackage,
+                handlers,
+                orderedHandlers,
+                (errorState) -> recordErrorState(errorState),
+                new ActionListener<>() {
+                    @Override
+                    public void onResponse(ActionResponse.Empty empty) {
+                        logger.info("Successfully applied new cluster state for namespace [{}]", namespace);
+                        errorListener.accept(null);
+                    }
+
+                    @Override
+                    public void onFailure(Exception e) {
+                        logger.error("Failed to apply immutable cluster state", e);
+                        errorListener.accept(e);
+                    }
+                }
+            ),
+            ClusterStateTaskConfig.build(Priority.URGENT),
+            new ImmutableStateUpdateStateTask.ImmutableUpdateStateTaskExecutor(namespace, clusterService.getRerouteService())
```

Comment:
Using a new executor instance each time will mean that these tasks are unbatched. Probably not a big deal in this context but as a rule we should re-use executor instances so that state updates are always batched.